Use different return code when some image lists aren't successfully processed

When image list cannot be processed (unavailable server etc.) cloudkeeper logs this issue and continues with rest of the image lists. In other words this is not considered an error and return code for this case is 0. Cloudkeeper should return different return code to signalize the issue.

I catch the same issue with an unavailable server:
2018-04-04T14:22:01+02:00 [ERROR] 36957 : Unexpected error: Connection timed out - connect(2) for "155.210.198.193" port 80
/opt/cloudkeeper/embedded/lib/ruby/2.2.0/net/http.rb:879:in `initialize': Connection timed out - connect(2) for "155.210.198.193" port 80 (Errno::ETIMEDOUT)
	from /opt/cloudkeeper/embedded/lib/ruby/2.2.0/net/http.rb:879:in `open'
	from /opt/cloudkeeper/embedded/lib/ruby/2.2.0/net/http.rb:879:in `block in connect'
	from /opt/cloudkeeper/embedded/lib/ruby/2.2.0/timeout.rb:73:in `timeout'

In this case, the following image lists are not processed.
FYI, I have sent an email to the image owner to tell that the link is not working, but I have still no answer. We must ignore such errors.

--- a/lib/cloudkeeper/backend_connector.rb
+++ b/lib/cloudkeeper/backend_connector.rb
@@ -2,11 +2,12 @@ module Cloudkeeper
   class BackendConnector
     include Cloudkeeper::Entities::Conversions
 
-    attr_reader :grpc_client, :nginx
+    attr_reader :grpc_client, :nginx, :errors
 
     def initialize
       @grpc_client = CloudkeeperGrpc::Communicator::Stub.new(Cloudkeeper::Settings[:'backend-endpoint'], credentials)
       @nginx = Cloudkeeper::Nginx::HttpServer.new
+      @errors = false
     end
 
     def pre_action
@@ -87,6 +88,7 @@ module Cloudkeeper
       errors = CloudkeeperGrpc::Constants.constants.reduce({}) { |acc, el| acc.merge(CloudkeeperGrpc::Constants.const_get(el) => el) }
       message = "#{errors[ex.code]}: #{ex.details}"
       logger.error "Backend error: #{message}"
+      @errors = true
       raise Cloudkeeper::Errors::BackendError, message if exception
       default
     end

--- a/lib/cloudkeeper/cli.rb
+++ b/lib/cloudkeeper/cli.rb
@@ -3,6 +3,9 @@ require 'yell'
 
 module Cloudkeeper
   class CLI < Thor
+    IMAGE_LIST_ERROR_CODE = 11
+    BACKEND_ERROR_CODE = 12
+
     class_option :'logging-level',
                  required: true,
                  default: Cloudkeeper::Settings['logging']['level'],
@@ -129,7 +132,7 @@ module Cloudkeeper
       initialize_sync options
       File.open(Cloudkeeper::Settings[:'lock-file'], File::RDWR | File::CREAT, 0o644) do |file|
         lock = file.flock(File::LOCK_EX | File::LOCK_NB)
-        Cloudkeeper::Managers::ApplianceManager.new.synchronize_appliances if lock
+        run_sync if lock
         abort 'cloudkeeper instance is already running, quitting' unless lock
       end
     rescue Cloudkeeper::Errors::InvalidConfigurationError => ex
@@ -147,6 +150,19 @@ module Cloudkeeper
     default_task :sync
 
     private
+
+    def run_sync
+      appliance_manager = Cloudkeeper::Managers::ApplianceManager.new
+      appliance_manager.synchronize_appliances
+
+      exit_with_message IMAGE_LIST_ERROR_CODE if appliance_manager.errors[:image_list_errors]
+      exit_with_message BACKEND_ERROR_CODE if appliance_manager.errors[:backend_errors]
+    end
+
+    def exit_with_message(code)
+      warn 'Some errors occured during the run. See logs for more info.'
+      exit code
+    end
 
     def initialize_sync(options)
       initialize_configuration options

--- a/lib/cloudkeeper/managers/appliance_manager.rb
+++ b/lib/cloudkeeper/managers/appliance_manager.rb
@@ -27,6 +27,10 @@ module Cloudkeeper
         abort ex.message
       end
 
+      def errors
+        { backend_errors: backend_connector.errors, image_list_errors: image_list_manager.errors }
+      end
+
       private
 
       def sync_expired_image_lists

--- a/lib/cloudkeeper/managers/image_list_manager.rb
+++ b/lib/cloudkeeper/managers/image_list_manager.rb
@@ -6,10 +6,11 @@ require 'json'
 module Cloudkeeper
   module Managers
     class ImageListManager
-      attr_reader :image_lists, :openssl_store
+      attr_reader :image_lists, :openssl_store, :errors
 
       def initialize
         @image_lists = {}
+        @errors = false
 
         @openssl_store = OpenSSL::X509::Store.new
         @openssl_store.add_path Cloudkeeper::Settings[:'ca-dir'] if Cloudkeeper::Settings[:'ca-dir']
@@ -33,6 +34,7 @@ module Cloudkeeper
           rescue Cloudkeeper::Errors::ImageList::DownloadError, Cloudkeeper::Errors::ImageList::VerificationError,
                  Cloudkeeper::Errors::Parsing::ParsingError => ex
             logger.warn "Image list #{url} couldn't be downloaded\n#{ex.message}"
+            @errors = true
             next
           end
         end

